backend/src/apis/fetch.rs

```diff
@@ -1,151 +1,102 @@
-        filter_balances_by_token_contracts(asset_contract_addresses, &res).unwrap();
-
-    Ok(filtered_balances)
-}
-
-/// This function filters out only the balances corresponding to an asset listed in asset_contract_addresses.
-fn filter_balances_by_token_contracts(
-    asset_contract_addresses: Vec<String>,
-    response: &Response,
-) -> Result<Vec<u64>, &'static str> {
-    let mut balances = Vec::new();
-    for contract_adddress in asset_contract_addresses {
-        if let Some(item) = response.data.items.iter().find(|&item| {
-            item.contract_address.to_ascii_lowercase() == contract_adddress.to_ascii_lowercase()
-        }) {
-            match item.balance.parse::<u64>() {
-                Ok(num) => balances.push(num),
-                Err(e) => println!("Failed to parse string: {}", e),
-            }
-        } else {
-            balances.push(0 as u64);
+pub async fn fetch_asset_sums<'a, M: Middleware + 'a>(
+    client: Arc<SignerMiddleware<Provider<Http>, LocalWallet>>,
+    contracts: Vec<Box<dyn BalanceFromContract<M> + Send>>,
+    asset_addresses: Vec<H160>,
+) -> Result<Vec<U256>, Box<dyn Error>> {
+    // For checking connectivity
+    client.get_block_number().await?;
+
+    let mut result: Vec<U256> = Vec::new();
+
+    let mut get_balance_futures = Vec::new();
+    for addr in asset_addresses.clone() {
+        get_balance_futures.push(client.get_balance(addr, None));
+    }
+    let get_balances = try_join_all(get_balance_futures).await?;
+    let sum_eth_balance = get_balances
+        .into_iter()
+        .reduce(|acc, balance| acc + balance)
+        .unwrap();
+    result.push(sum_eth_balance);
+
+    // Most in case, the number of contracts is less than the number of asset addresses
+    // Iterating over contracts first is more efficient
+    let mut sum_erc_balance = U256::zero();
+    for contract in &contracts {
+        for addr in asset_addresses.clone() {
+            let erc_balance = contract
+                .get_balance_from_contract(addr)
+                .call()
+                .await
+                .unwrap();
+            sum_erc_balance = sum_erc_balance + erc_balance;
         }
+        result.push(sum_erc_balance)
     }
-    Ok(balances)
+
+    Ok(result)
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
 
-    #[test]
-    #[ignore]
-    fn test_fetch_balances_per_addr() {
-        // this is an address with 0 usdc and 0.010910762665574143 ETH
-        let address = "0xe4D9621321e77B499392801d08Ed68Ec5175f204".to_string();
+    use crate::contracts::tests::initialize_anvil;
 
-        let usdc_contract = "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48".to_string();
-        let eth_contract = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee".to_string();
+    use ethers::utils::Anvil;
 
-        let balances =
-            fetch_balances_per_addr(address, [usdc_contract, eth_contract].to_vec()).unwrap();
+    #[tokio::test]
+    async fn test_fetch_asset_sums() {
```

Comment:
I don't understand this test. Are the 2 `asset_sums` returned related to 2 different erc20 contracts. Or is it one `eth` and the other `mockerc20` ?

If so how are you able to set the balance of the `cex_addr_1` and `cex_addr_2` inside the erc20?

Reply:
> I don't understand this test. Are the 2 `asset_sums` returned related to 2 different erc20 contracts. Or is it one `eth` and the other `mockerc20` ?

It returns the balance of both ETH and a mock ERC20 token.

> If so how are you able to set the balance of the `cex_addr_1` and `cex_addr_2` inside the erc20?

the assets are configured within the [`initialize_anvil`](https://github.com/summa-dev/summa-solvency/blob/abf920fccc4818a43fce00a142b4a5ba6c12c82a/backend/src/contracts/tests.rs#L12-L46) function

https://github.com/summa-dev/summa-solvency/blob/abf920fccc4818a43fce00a142b4a5ba6c12c82a/backend/src/contracts/tests.rs#L12-L46

For the purpose of this test, you can think of the test chain as `Anvil` instead of `Hardhat`